### ISSUE TITLE
Basic how-to for embedding a new language (with coroutines)

### ISSUE THREAD

**Issue by rockobonaparte** [IronLanguages/dlr#171]
I can't find any good (up-to-date, anyways) documentation for creating and embedding a new language using the DLR. Since the GitHub project recommended just opening issue, I'm being shameless and submitting it here. I am trying to specify a domain-specific language for NPC and game logic for a hobby game in Unity 3d. So I was particularly trying to define a syntax that worked well using green threads and coroutines. But right now I can't even figure out how to run a calculator.

I could just post some code if desired in multiple layers of disclosure since it's just a trial project, but I'm away from it right now. Eventually I think I'd just pastebin the whole mess since I think the important parts are just in a single file. I defined a grammar using ANTLR4 and I have a visitor that is able to crunch a one-liner calculator-style program statement. My goal was to then make that play in the DLR. Here's what I have so far:

1. I create a ScriptRuntime for my language
2. I get ScriptEngine for my language from that ScripRuntime
3. I create a ScriptSource using that engine using a one-line string for my test program
   * ...at this point, it's going to my LanguageContext.CompileSourceCode implementation, which is where I use my ANTLR4 stuff to make a System.Linq.Expression root node. I shove this into a ScriptCode subclass I created
   * ...something inside the DLR engine internals knows to call the Run() method on my custom ScriptCode, which is where I run my expression as a lambda

The problem is that it's just an incestuous loop right now. I mean, it does actually work, but it's just a bunch of handling to do what I already did. I assume instead of System.Linq.Expression that I should be turning this into some kind of byte code and feeding it to some interpreter. I see an interpreter:

https://github.com/IronLanguages/dlr/blob/master/Src/Microsoft.Dynamic/Interpreter/Interpreter.cs

And the bulk of its syntax:
https://github.com/IronLanguages/dlr/tree/master/Src/Microsoft.Dynamic/Interpreter/Instructions

Is this what the cool kids are using? What's the general route where I get that thing going and I feed it my stuff?

All the information on this is too old and everything seems to have changed. The iron language projects are sitting under many, many layers and I lose track of what is going on. I can't figure out what I should be doing from that--especially for something basic for starters.

I also wonder then how I am supposed to get debug capability, as I believe that is one perk of the DLR. Do I need to associate PDB data with my script or something? Where do I pump that in?

The other issue is after all this, I don't even know if that interpreter is going to help me with green threads. I would want to trip across a custom instruction that would tell the interpreter to stop. At that point, I want to be able to take the entire context the interpreter was using, store it into a frame, and possibly just move on to a totally different script. Preferably I could even serialize that frame so I can save and load it from game data. I'd rather have that than create chains of yields and things like that. Preferably, my parser/pre-compiler would tactfully insert these before major calls and such too.

**Comment by rockobonaparte** [comment]
FWIW if I could add a new interpreter instruction to halt the interpreter to change frames and basically create ironStacklessPython then that might be easy enough and do what I want. A lot of what is motivating this is that yield would force me to write yield, like, everywhere; I don't think IronPython 3 has a good handle on asyncio and I wouldn't even be using it in a normal way.

**Comment by slozier** [comment]
I'm not sure if it's of any use (I only know that it exists), but have you looked at the sympl example (Samples/sympl/csharp)? There's also an associated doc (Docs/sympl.pdf) which explains how it is implemented.

**Comment by rockobonaparte** [comment]
I looked at that too and I think it was the inspiration for what I originally saw for using System.Linq.Expressions. Unfortunately, I can't use coroutines with those--I can't even yield if I understand lambdas in System.Linq. I thing I ran into the the simple interpreter in Microsoft.Scripting before and couldn't figure out how I might preempt it. So it sounds like it's one of those things I'd have to roll myself if I want it that much. :(
